contracts/sg-ics721/src/execute.rs

```diff
@@ -113,4 +115,70 @@ impl Ics721Execute for SgIcs721Contract {
 
         to_json_binary(&instantiate_msg)
     }
+
+    fn mint_msg(
+        &self,
+        token_id: String,
+        token_uri: Option<String>,
+        owner: String,
+        data: Option<Binary>,
+    ) -> StdResult<Binary> {
+        // parse token data and check whether it is of type NftExtension
+        let extension: Metadata = match data {
+            Some(data) => {
+                match from_json::<NftExtension>(data).ok().map(|ext| Metadata {
+                    animation_url: ext.animation_url,
+                    attributes: ext.attributes.map(|traits| {
+                        traits
+                            .into_iter()
+                            .map(|t| Trait {
+                                trait_type: t.trait_type,
+                                value: t.value,
+                                display_type: t.display_type,
+                            })
+                            .collect()
+                    }),
+                    background_color: ext.background_color,
+                    description: ext.description,
+                    external_url: ext.external_url,
+                    image: ext.image,
+                    image_data: ext.image_data,
+                    youtube_url: ext.youtube_url,
+                    name: ext.name,
+                }) {
+                    Some(extension) => extension,
+                    None => Metadata {
```

Comment:
can you pre-build this empty Metadata and then insert it so you don't have to write it out twice?

Reply:
Changed and simplified code:

```rust
        let extension = data
            .and_then(|binary| {
                from_json::<NftExtension>(binary).ok().map(|ext| Metadata {
                    animation_url: ext.animation_url,
                    attributes: ext.attributes.map(|traits| {
                        traits
                            .into_iter()
                            .map(|t| Trait {
                                trait_type: t.trait_type,
                                value: t.value,
                                display_type: t.display_type,
                            })
                            .collect()
                    }),
                    background_color: ext.background_color,
                    description: ext.description,
                    external_url: ext.external_url,
                    image: ext.image,
                    image_data: ext.image_data,
                    youtube_url: ext.youtube_url,
                    name: ext.name,
                })
            })
            .unwrap_or(Metadata {
                // no onchain metadata (only offchain), in this case empty metadata is created
                ..Default::default()
            });
```